transport/tracer_test.go

```diff
@@ -0,0 +1,295 @@
+		Tracer: tracer,
+	})
+	assert.NoError(t, err)
+	hp := "127.0.0.1:4040"
+	ch.ListenAndServe(hp)
+
+	rpc := yarpc.NewDispatcher(yarpc.Config{
+		Name: "yarpc-test",
+		Inbounds: []transport.Inbound{
+			ytchannel.NewInbound(ch),
+		},
+		Outbounds: transport.Outbounds{
+			"yarpc-test": ytchannel.NewOutbound(ch, ytchannel.HostPort(hp)),
+		},
+		Tracer: tracer,
+	})
+
+	json.Register(rpc, json.Procedure("echo", echo))
+
+	client := json.New(rpc.Channel("yarpc-test"))
+
+	rpc.Start()
+	defer rpc.Stop()
+
+	ctx := context.Background()
+	ctx, cancel := context.WithTimeout(ctx, 100*time.Millisecond)
+	defer cancel()
+
+	var resBody echoResBody
+	_, err = client.Call(
+		ctx,
+		yarpc.NewReqMeta().Procedure("echo"),
+		&echoReqBody{},
+		&resBody,
+	)
+	assert.NoError(t, err)
+
+	AssertDepth1Spans(t, tracer)
+}
+
+func TestHttpTracerDepth2(t *testing.T) {
+	tracer := mocktracer.New()
+
+	ctx := context.Background()
+	ctx, cancel := context.WithTimeout(ctx, 100*time.Millisecond)
+	defer cancel()
+
+	span := tracer.StartSpan("test")
+	// no defer span.Finish()
+	span.SetBaggageItem("weapon", "knife")
+	ctx = opentracing.ContextWithSpan(ctx, span)
+
+	rpc := yarpc.NewDispatcher(yarpc.Config{
+		Name: "yarpc-test",
+		Inbounds: []transport.Inbound{
+			http.NewInbound(":8080"),
+		},
+		Outbounds: transport.Outbounds{
+			"yarpc-test": http.NewOutbound("http://127.0.0.1:8080"),
+		},
+		Tracer: tracer,
+	})
+
+	client := json.New(rpc.Channel("yarpc-test"))
+
+	echo := func(ctx context.Context, reqMeta yarpc.ReqMeta, reqBody *echoReqBody) (*echoResBody, yarpc.ResMeta, error) {
+		span := opentracing.SpanFromContext(ctx)
+		weapon := span.BaggageItem("weapon")
+		assert.Equal(t, "knife", weapon, "baggage should propagate")
+		return &echoResBody{}, nil, nil
+	}
+
+	echoecho := func(ctx context.Context, reqMeta yarpc.ReqMeta, reqBody *echoReqBody) (*echoResBody, yarpc.ResMeta, error) {
+		var resBody echoResBody
+		_, err := client.Call(
+			ctx,
+			yarpc.NewReqMeta().Procedure("echo"),
+			reqBody,
+			&resBody,
+		)
+		if err != nil {
+			return nil, nil, err
+		}
+		return &resBody, nil, nil
+	}
+
+	json.Register(rpc, json.Procedure("echo", echo))
+	json.Register(rpc, json.Procedure("echoecho", echoecho))
+
+	rpc.Start()
+	defer rpc.Stop()
+
+	var resBody echoResBody
+	_, err := client.Call(
+		ctx,
+		yarpc.NewReqMeta().Procedure("echoecho"),
+		&echoReqBody{},
+		&resBody,
+	)
+	assert.NoError(t, err)
+
+	assert.Equal(t, 4, len(tracer.FinishedSpans()), "generates inbound and outband spans")
+	if len(tracer.FinishedSpans()) != 4 {
+		return
+	}
+	spans := tracer.FinishedSpans()
+	assert.Equal(t, "echo", spans[0].OperationName, "span has correct operation name")
+	assert.Equal(t, "echo", spans[1].OperationName, "span has correct operation name")
+	assert.Equal(t, "echoecho", spans[2].OperationName, "span has correct operation name")
+	assert.Equal(t, "echoecho", spans[3].OperationName, "span has correct operation name")
```

Comment:
this needs to check that trace ID is the same across all spans. If it's not the same, the propagation is actually broken somewhere, so even though the spans are reported, the trace itself is broken.